README.md
## Famo.us Carousel

A responsive carousel / slideshow / slider powered by Famo.us.  Supports mouse, touch/swipe and keyboard navigation.<BR>

***
__UPDATE: 0.9.x introduces ES6 and a new require() path:__

CommonJS:

        var famousCarousel = require("famous-carousel").Carousel;
        var carousel = new famousCarousel( {...} );

Global:

        var carousel = famousCarousel.default( {...} );

__If you see any unusual behavior after upgrading, delete node\_modules and run _npm install_ again__
<BR>
***

![famous-carousel preview](https://cloud.githubusercontent.com/assets/6295083/8266024/78a6db1c-16dc-11e5-9c18-93a25824a72c.gif)

After cloning the repo:

    npm install

Start the example project at port 8080:

    npm run example

### Usage

famous-carousel uses Browserify require. Refer to the _example_ folder for boilerplate setup. When using this as an npm module, reference it as such:

    var Carousel = require("famous-carousel").Carousel;
    var myCarousel = new Carousel("myDivSelector", {
                            // add options here
                            // ...
                        }
                    );

  ####ES6 is now supported
    import { Carousel } from "famous-carousel";

### Carousel Options
The following keys are supported in the options object. Only _carouselData_ is required, all others are optional.

* #### _selector_
Type: `String` or `Object`<BR>
Default: `body`<BR>
As a string, _selector_ is a CSS selector of the element to render into.<BR>
As an object, _selector_ is assumed to be a Famous node.

* #### carouselData (required)
Type: `Array`<BR>
This specifies the content of the slides. It is an array of objects, each containing _type_ and _data_ keys.<BR>
_type_ may be `image`, `markup`, or `node`.<BR>
_data_ must be a url for image, any valid html for markup, or a Famo.us node object.<BR>
Example data:<BR>
        [
            {   "type": "image",
                "data": "http://myDomain/myPicture.jpg"
            },
            {   "type": "markup",
                "data": "<div style='color:blue'>Hello World<BR><BR>It's me!</div>"
            },
            {   "type": "node",
                "data": myFamousNode
            }
        ]

* #### initialIndex
Type: `Integer`<BR>
Default: `0`<BR>
Initial slide number to show (zero-based). If larger than the number of slides, it will cap to the final slide.

* #### wrapAround
Type: `Boolean`<BR>
Default: `false`<BR>
By default, navigation stops at the ends of the slide set. Setting wrapAround to true allows navigation to wrap from one end to the other.

* #### autoPlay
Type: `Integer`<BR>
Default: `0`<BR>
Automatically transition slides, pausing on each slide for the number of milliseconds specified. When the user navigates manually, automatic playback stops.

* #### autoSlidesToAdvance
Type: `Integer`<BR>
Default: `1`<BR>
Number of slides to advance during automatic playback.

* #### manualSlidesToAdvance
Type: `Integer`<BR>
Default: `1`<BR>
Number of slides to advance when user navigates.

* #### dotWidth
Type: `Integer`<BR>
Default: `10`<BR>
Width of dots along bottom.

* #### dotSpacing
Type: `Integer`<BR>
Default: `5`<BR>
Spacing of dots along bottom.

* #### dotForeColor
Type: `String`<BR>
Default: `"white"`<BR>
Navigation dot foreground color (CSS style).

* #### dotBackColor
Type: `String`<BR>
Default: `"transparent"`<BR>
Navigation dot background color (CSS style).

* #### dotOpacity
Type: `Float`<BR>
Default: `1.0`<BR>
Navigation dot opacity (CSS style).

* #### dotSelectedClass
Type: `String`<BR>
CSS class to style the selected dot. This overrides the other dot* CSS options (dotWidth & dotSpacing continue to be honored).

* #### dotUnselectedClass
Type: `String`<BR>
CSS class to style unselected dots. This overrides the other dot* CSS options (dotWidth & dotSpacing continue to be honored).

* #### arrowFillColor
Type: `String`<BR>
Default: `"white"`<BR>
Navigation arrow fill color (CSS style).

* #### arrowOutlineColor
Type: `String`<BR>
Default: `"transparent"`<BR>
Navigation arrow outline color (CSS style).

* #### arrowClass
Type: `String`<BR>
CSS class to style dots. This overrides the other arrow* CSS options.


### Public methods
* #### updateData()
Updates the slide data. This calls clearSlides() to reinitialize the slides. There is an issue with the dots not re-rendering properly.

* #### clearSlides()
Removes the slides.

* #### clearAll()
Removes the carousel instance entirely. NOTE: Famo.us 0.5.2 has [a bug causing it not to remove the DOMElement](https://github.com/Famous/engine/issues/245). famous-carousel removes the div.famous-dom-renderer element for now.<BR>

### Building
To build a self-contained bundles:

    $ npm run build

Builds
 - CommonJS version (dist/famous-carousel.min.js)
 - global version (dist/global/famous-carousel.min.js), and global debug version (dist/global/famous-carousel.debug.js).  
 The global build uses the variable name _famousCarousel_.  
 To see the global version in action open browser to local port 8080/_global.html_ after:  

      $ npm run example

To build optional es5 code:

     $ npm install -g babel
     $ npm run build-es5

Run tests (linter & style checks for now):

    npm run test


### FAQ
**Q:** The arrows & dots show, but the slides are invisible.<BR>
**A:** famous-carousel sizes to its container's dimensions. Due to CSS box model rules, if no height is specified for the container, the default auto height computes to 0. Short answer: set a height for its parents all the way up to the root element.

**Q:** The slides appear from outside the containing element when animating.<BR>
**A:** Set the container's overflow property to _hidden_. Additionally center the famous-carousel within the containing element using these properties:

    .myContainerClass {
        overflow:hidden;
        position:absolute;
        left:0; right:0;
        top:0; bottom:0;
        margin:auto;
    }

**Q:** The arrows/dots aren't visible [against a white backgound].<BR>
**A:** _dotBackColor_ and _arrowOutlineColor_ default to transparent. Set them to a different color.

**Q:** The slides are shifted down and to the right.<BR>
**A:** The selector element must not have any padding. The work-around for this [Famous engine bug](https://github.com/Famous/engine/issues/411) is to wrap the selector element in a container element which has the padding. Remove all padding from the selector element. See _example/index.html_ for... an example.

### To Do
Pull requests are welcome. When submitting a PR, please make sure _npm run test_ passes.
* Unit tests.
* Add exit transition.
* Add more user-configurable options such as slide transitions, nav arrow images...
* Option to auto-hide navigation arrows & dots.
* Navigate by clicking on dots.
* Auto start/stop videos in slides.
* Option for vertical scrolling.
* Support CSS class for slides?

### License
This project is based on the [Famo.us Carousel tutorial](http://famous.org/learn/Carousel/index.html) and bound by the same MIT license. Refer to the LICENSE file.


CHANGELOG.md
v0.9.0 - June 25, 2015

* Update to Famous engine 0.6.2 to avoid npm issue with 0.5.x.
* Fix events to match changes in 0.6.x's eventing API.
* BREAKING: merged talves' ES6 module support.
		// Browserify usage:
		var famousCarousel = require("famous-carousel").Carousel;
		var carousel = new famousCarousel( {...} );

v0.8.2 - June 21, 2015

* Support Famo.us nodes in carouselData.

v0.8.1 - June 20, 2015

* Support passing in a Famous node as the selector.

v0.8.0 - June 19, 2015

* Initial release based on the Famous carousel tutorial.


src/Pager.js
/**
 * Pager.js
 */

import DOMElement from "famous/dom-renderables/DOMElement";
import PhysicsEngine from "famous/physics/PhysicsEngine";
import FamousEngine from "famous/core/FamousEngine";
import GestureHandler from "famous/components/GestureHandler";

import physics from "famous/physics";
import math from "famous/math";

var Box = physics.Box;
var Spring = physics.Spring;
var Vec3 = math.Vec3;
var hideAngle = Math.PI / 2;

export class Pager {
	constructor(options) {
		this.options = options;
		this.node = options.parent.addChild();
		this.currentIndex = options.initialIndex || 0;
		this.threshold = 4000;
		this.force = new Vec3();
		this.pageWidth = 0;

		// Add a physics simulation and update this instance using regular time updates from the clock.
		this.simulation = new PhysicsEngine();

		var resizeComponent = {
			onSizeChange: x => this.pageWidth = x
		};
		this.node.addComponent(resizeComponent);

		// .requestUpdate will call the .onUpdate method next frame, passing in the time stamp for that frame
		FamousEngine.requestUpdate(this);

		this.createPages(this.options);
	}

	onUpdate(time) {
		if (!this.node) {
			return;
		}

		this.simulation.update(time);

		var page;
		var physicsTransform;
		var p, r, xPos;
		for (var i = 0, len = this.pages.length; i < len; i++) {
			page = this.pages[i];

			// Get the transform from the `Box` body
			physicsTransform = this.simulation.getTransform(page.box);
			p = physicsTransform.position;

			// Set the `slide`s x-position to the `Box` body's x-position
			// Math.round is significantly faster than toFixed.
			xPos = Math.round(p[0] * this.pageWidth * 100) / 100;
			page.node.setPosition(xPos, 0, 0);

			// Set the currently selected page's rotation to match the `Box` body's rotation
			if (i === this.currentIndex) {
				r = physicsTransform.rotation;
				page.node.setRotation(r[0], r[1], r[2], r[3]);
			}
		}

		FamousEngine.requestUpdateOnNextTick(this);
	}

	pageChange(oldIndex, newIndex) {
		var xOffset;
		var reshuffle = false;
		// Wrap-around cases shuffle pages to other side.
		if (oldIndex === this.pages.length - 1 && newIndex === 0) {
			xOffset = 1;
			reshuffle = true;
		} else if (oldIndex === 0 && newIndex === this.pages.length - 1) {
			xOffset = -1;
			reshuffle = true;
		}

		if (reshuffle) {
			for (var i = 0; i < this.pages.length; i++) {
				this.pages[i].anchor.set(xOffset, 0, 0);
				this.pages[i].node.setRotation(0, hideAngle, 0);
			}
		} else {
			// Non-wrapping cases
			xOffset = (oldIndex < newIndex) ? -1 : 1;
			this.pages[oldIndex].anchor.set(xOffset, 0, 0);
			this.pages[oldIndex].node.setRotation(0, hideAngle, 0);
		}

		this.pages[newIndex].anchor.set(0, 0, 0);
		this.pages[newIndex].node.setRotation(0, 0, 0);
		this.currentIndex = newIndex;
	}

	removePages() {
		for (var i = 0; i < this.pages.length; i++) {
			//DOMElement bug https://github.com/Famous/engine/issues/245
			if (this.pages[i].el) {
				this.pages[i].el.setProperty("display", "none");
			}
			this.options.parent.removeChild(this.pages[i].node);
		}
		this.pages = [];
	}

	createPages(options) {
		var pages = [];

		for (var i = 0; i < options.carouselData.length; i++) {
			var slide, el;

			if (options.carouselData[i].type === "node") {
				slide = options.carouselData[i].data;
			} else {
				slide = this.node.addChild();
				el = new DOMElement(slide);
				switch (options.carouselData[i].type) {
					case "image":
						el.setProperty("backgroundImage", "url(" + options.carouselData[i].data + ")");
						el.setProperty("background-repeat", "no-repeat");
						el.setProperty("background-size", "contain");
						el.setProperty("background-position", "center");
						break;
					case "markup":
						el.setContent(options.carouselData[i].data);
						break;
				}
			}

			slide.setAlign(0.5, 0.5);
			slide.setMountPoint(0.5, 0.5);
			slide.setOrigin(0.5, 0.5);

			var gestureHandler = new GestureHandler(slide);
			/*eslint-disable */
			gestureHandler.on("drag", function(index, e) {
				this.force.set(e.centerDelta.x, 0, 0); // Add a force equal to change in X direction
				this.force.scale(20); // Scale the force up
				this.pages[index].box.applyForce(this.force); // Apply the force to the `Box` body

				var direction = 0;
				if (e.centerVelocity.x > this.threshold) {
					if (this.draggedIndex === index && this.currentIndex === index) {
						// Move index to left
						direction = -1;
					}
				} else if (e.centerVelocity.x < -this.threshold) {
					if (this.draggedIndex === index && this.currentIndex === index) {
						direction = 1;
					}
				}
				if (direction !== 0) {
					this.options.context.emit("pageChange", {
						direction: direction,
						numSlidesToAdvance: this.options.manualSlidesToAdvance,
						stopAutoPlay: true
					});
				}

				if (e.status === "start") {
					this.draggedIndex = index;
				}
			}.bind(this, i));
			/*eslint-enable */

			// A `Box` body to relay simulation data back to the visual element
			var box = new Box({
				mass: 100,
				size: [100, 100, 100]
			});

			// Place all anchors off the screen & rotated, except for the
			// anchor belonging to the first/initial image node
			var anchor = new Vec3(1, 0, 0); //off-screen
			var yAngle = hideAngle;
			if (i === this.currentIndex) {
				anchor = new Vec3(0, 0, 0); //on-screen
				yAngle = 0;
			}
			slide.setRotation(0, yAngle, 0);

			// Attach the box to the anchor with a `Spring` force
			//ToDo: Add support for user-configurable transition
			var spring = new Spring(null, box, {
				period: 0.3,
				dampingRatio: 0.7,
				anchor: anchor
			});

			// Notify the physics engine to track the box and the springs
			this.simulation.add(box, spring);

			pages.push({
				node: slide,
				el: el,
				box: box,
				spring: spring,
				anchor: anchor
			});
		}

		this.pages = pages;
	}

}
